keps/sig-node/2008-checkpoint-restore/README.md

```diff
@@ -0,0 +1,320 @@
+   In addition to the simplest form of container migration (checkpoint,
+   transfer, restore) LXD also supports optimizations to decrease the
+   downtime during migration by using CRIU's pre-copy migration support.
+
+   * https://archive.fosdem.org/2018/schedule/event/containers_optimized_migration/
+   * https://lisas.de/~adrian/posts/2017-Dec-06-optimizing-live-container-migration-in-lxd.html
+
+ * Borg uses CRIU to live migrate containers between hosts to free up resources
+   on hosts which are missing resources under load:
+
+   * [Task Migration at Google Using CRIU](https://www.linuxplumbersconf.org/event/2/contributions/209/)
+   * [Update on Task Migration at Google Using CRIU](https://linuxplumbersconf.org/event/4/contributions/508/)
+
+   See especially the second presentation for limitations Google is
+   experiencing using CRIU based container migration in production.
+
+ * Podman supports container migration in its simplest form (checkpoint,
+   transfer, restore): https://criu.org/Podman
+
+The motivation is to get closer to container migration by taking the first
+step and providing support for simple checkpoint and restore. The motivation
+is existing checkpoint/restore/migration support in other container
+environments showing that it is useful and production ready.
+
+### Goals
+
+The goal of this KEP is to introduce *checkpoint* and *restore* to the CRI API.
+
+This includes extending the *kubelet* to support checkpointing single containers
+with the User Story 1 in mind (forensic use case).
+
+### Non-Goals
+
+Out of scope of this KEP are high level discussions about how to implement
+container migration. This is only about low level primitives to add
+*checkpoint* and *restore* to the API.
+
+## Proposal
+
+### Implementation
+
+#### Phase 1
+
+One of the first goals is to implement User Story 1. For this forensic
+use case we want to offer the possibility to checkpoint a container
+out of a running Pod without stopping the checkpointed container and without
+letting the container know that it was checkpointed.
+
+The corresponding code changes for the forensic use case can be found in the
+following pull request:
+
+* https://github.com/kubernetes/kubernetes/pull/104907
+
+The goal is to introduce *checkpoint* and *restore* in a bottom-up approach.
+In a first step we only want to extend the CRI API to trigger a checkpoint
+by the container engine and to have the low level primitives in the *kubelet*
+to trigger a checkpoint.
+
+In the corresponding pull request a checkpoint is triggered using the *kubelet*
+API:
+
+```
+curl -skv -X POST "https://localhost:10250/checkpoint/default/counters/wildfly"
+```
+
+For the first implementation we do not want to support restore in the
+*kubelet*. With the focus on the forensic use case the restore should happen
+outside of Kubernetes. The restore is a container engine only operation
+in this first step.
+
+The forensic use case is targeted to be part of the next (1.23) release.
```

Comment:
We should also cover adding a feature flag for this.

Reply:
done